File: web/packages/alloy-audio/src/synth-engine-core.ts
// The polyphony + sustain-pedal state machine, decoupled from tone production
// via VoicePlayer and from platform time via now(). Main-thread-only by
// design; the platform wrapper owns cross-thread hand-off (and, on the web,
// autoplay resume). Web-canonical twin of AlloyAudio's SynthEngineCore.swift.

import type { SynthEngine } from './synth-engine.js';
import type { ActiveVoice, VoicePlayer } from './voice-player.js';

interface Voice {
  active: ActiveVoice;
  heldByKey: boolean; // key is still physically down
  heldByPedal: boolean; // sustain pedal latched it
}

export class SynthEngineCore implements SynthEngine {
  private readonly voices = new Map<number, Voice>();
  private sustain = false;
  // No instrument selected yet: noteOn is a no-op until the platform wrapper
  // (or the app) calls setInstrument with its default id.
  private player: VoicePlayer | null = null;

  constructor(
    private readonly playerFor: (id: string) => VoicePlayer,
    private readonly now: () => number,
  ) {}

  noteOn(midi: number, velocity = 1): void {
    if (!this.player) {
      return;
    }
    const existing = this.voices.get(midi);
    if (existing) {
      // Already sounding; the envelope is intentionally not re-struck. But re-assert
      // the physical hold and clear any pedal latch left by a prior release, so a
      // later pedal-up does not release a key that is still physically down.
      existing.heldByKey = true;
      existing.heldByPedal = false;
      return;
    }
    const active = this.player.start(midi, velocity, this.now());
    this.voices.set(midi, { active, heldByKey: true, heldByPedal: false });
  }

  noteOff(midi: number): void {
    const voice = this.voices.get(midi);
    if (!voice) {
      return;
    }
    voice.heldByKey = false;
    if (this.sustain) {
      voice.heldByPedal = true;
      return;
    }
    this.release(midi, voice);
  }

  setSustain(on: boolean): void {
    this.sustain = on;
    if (!on) {
      for (const [midi, voice] of [...this.voices]) {
        if (voice.heldByPedal && !voice.heldByKey) {
          this.release(midi, voice);
        }
      }
    }
  }

  setInstrument(id: string): void {
    this.player = this.playerFor(id);
  }

  allNotesOff(): void {
    const when = this.now();
    for (const [midi, voice] of [...this.voices]) {
      voice.active.stop(when);
      this.voices.delete(midi);
    }
  }

  private release(midi: number, voice: Voice): void {
    voice.active.release(this.now());
    this.voices.delete(midi);
  }
}

File: swift/Sources/AlloyAudio/SynthEngineCore.swift
/// Note/sustain state machine. Mirrored twin of the web alloy-audio
/// `engine-core.ts` (the canonical form of this shape).

/// One sounding note, owned by the engine's per-note voice map.
public protocol ActiveVoiceHandle {
    /// Begin the key-up release; the voice tears itself down when silent.
    func release(at when: Double)
    /// Hard stop (allNotesOff): fast fade, then teardown.
    func stop(at when: Double)
}

/// Tone-production strategy for one instrument.
public protocol VoicePlayer {
    func start(midi: Int, velocity: Double, at when: Double) -> ActiveVoiceHandle
}

/// The playing surface the UI talks to. Mirrors the web SynthEngine.
public protocol SynthEngine: AnyObject {
    func noteOn(midi: Int, velocity: Double)
    func noteOff(midi: Int)
    func setSustain(_ on: Bool)
    func setInstrument(_ id: String)
    func allNotesOff()
}

public extension SynthEngine {
    func noteOn(midi: Int) { noteOn(midi: midi, velocity: 1) }
}

/// The polyphony + sustain-pedal state machine, decoupled from tone
/// production via VoicePlayer. Main-thread-only by design; the platform
/// adapter owns cross-thread hand-off.
public final class SynthEngineCore: SynthEngine {
    private struct VoiceRecord {
        var active: ActiveVoiceHandle
        var heldByKey: Bool // key is still physically down
        var heldByPedal: Bool // sustain pedal latched it
    }

    private var voices: [Int: VoiceRecord] = [:]
    private var sustain = false
    // Nil until the wrapper (or the app) calls setInstrument with its default
    // id — same contract as the web twin: noteOn is a no-op with no player.
    private var player: VoicePlayer?
    private let playerFor: (String) -> VoicePlayer
    private let now: () -> Double

    public init(
        playerFor: @escaping (String) -> VoicePlayer,
        now: @escaping () -> Double,
    ) {
        self.playerFor = playerFor
        self.now = now
    }

    public func noteOn(midi: Int, velocity: Double) {
        guard let player else { return }
        if voices[midi] != nil {
            // Already sounding; the envelope is intentionally not re-struck.
            // Re-assert the physical hold and clear any pedal latch so a later
            // pedal-up does not release a key that is still physically down.
            voices[midi]?.heldByKey = true
            voices[midi]?.heldByPedal = false
            return
        }
        let active = player.start(midi: midi, velocity: velocity, at: now())
        voices[midi] = VoiceRecord(active: active, heldByKey: true, heldByPedal: false)
    }

    public func noteOff(midi: Int) {
        guard var voice = voices[midi] else { return }
        voice.heldByKey = false
        if sustain {
            voice.heldByPedal = true
            voices[midi] = voice
            return
        }
        release(midi: midi, voice: voice)
    }

    public func setSustain(_ on: Bool) {
        sustain = on
        guard !on else { return }
        for (midi, voice) in voices where voice.heldByPedal && !voice.heldByKey {
            release(midi: midi, voice: voice)
        }
    }

    public func setInstrument(_ id: String) {
        player = playerFor(id)
    }

    public func allNotesOff() {
        let when = now()
        for voice in voices.values {
            voice.active.stop(at: when)
        }
        voices.removeAll()
    }

    private func release(midi: Int, voice: VoiceRecord) {
        voice.active.release(at: now())
        voices.removeValue(forKey: midi)
    }
}
